src/System.Linq.Expressions/src/System/Dynamic/Utils/TypeUtils.cs

```diff
@@ -246,6 +246,70 @@ public static bool HasIdentityPrimitiveOrNullableConversionTo(this Type source,
             return false;
         }
 
+        public static bool IsValueTupleType(this Type type) =>
+            type.IsConstructedGenericType && IsValueTupleType(type, type.GetGenericArguments().Length);
+
+        private static bool IsValueTupleType(Type type, int argumentCount)
+        {
+            Debug.Assert(type.IsConstructedGenericType);
+            type = type.GetGenericTypeDefinition();
+            switch (argumentCount)
+            {
+                case 1:
+                    return type == typeof(ValueTuple<>);
+                case 2:
+                    return type == typeof(ValueTuple<,>);
+                case 3:
+                    return type == typeof(ValueTuple<,,>);
+                case 4:
+                    return type == typeof(ValueTuple<,,,>);
+                case 5:
+                    return type == typeof(ValueTuple<,,,,>);
+                case 6:
+                    return type == typeof(ValueTuple<,,,,,>);
+                case 7:
+                    return type == typeof(ValueTuple<,,,,,,>);
+                case 8:
+                    return type == typeof(ValueTuple<,,,,,,,>);
```

Comment:
A nonary or higher generic type could legitimately come here, and false would be the correct answer.